tests/test_shell_launcher.py

```diff
@@ -0,0 +1,323 @@
+    """Test that the popen call returns a popen object"""
+    run_dir, out_file, err_file = generate_directory(test_dir)
+    with (
+        open(out_file, "w", encoding="utf-8") as out,
+        open(err_file, "w", encoding="utf-8") as err,
+    ):
+        cmd = ShellLauncherCommand(
+            {},
+            run_dir,
+            subprocess.DEVNULL,
+            subprocess.DEVNULL,
+            EchoHelloWorldEntity().as_program_arguments(),
+        )
+        id = shell_launcher.start(cmd)
+    with shell_launcher._launched[id] as proc:
+        assert isinstance(proc, sp.Popen)
+
+
+def test_popen_writes_to_output_file(shell_launcher, test_dir: str):
+    """Test that popen writes to .out file upon successful process call"""
+    run_dir, out_file, err_file = generate_directory(test_dir)
+    with (
+        open(out_file, "w", encoding="utf-8") as out,
+        open(err_file, "w", encoding="utf-8") as err,
+    ):
+        cmd = ShellLauncherCommand(
+            {}, run_dir, out, err, EchoHelloWorldEntity().as_program_arguments()
+        )
+        id = shell_launcher.start(cmd)
+        out.close(), err.close()
+    proc = shell_launcher._launched[id]
+    assert proc.wait() == 0
+    assert proc.returncode == 0
+    with open(out_file, "r", encoding="utf-8") as out:
+        assert out.read() == "Hello World!\n"
+    with open(err_file, "r", encoding="utf-8") as err:
+        assert err.read() == ""
+
+
+@requires_slurm
+def test_popen_fails_with_invalid_cmd(shell_launcher, test_dir):
+    """Test that popen returns a non zero returncode after failure"""
+    run_dir, out_file, err_file = generate_directory(test_dir)
+    with (
+        open(out_file, "w", encoding="utf-8") as out,
+        open(err_file, "w", encoding="utf-8") as err,
+    ):
+        args = (helpers.expand_exe_path("srun"), "--flag_dne")
+        cmd = ShellLauncherCommand({}, run_dir, out, err, args)
+        id = shell_launcher.start(cmd)
+        proc = shell_launcher._launched[id]
+        proc.wait()
+        assert proc.returncode != 0
+        with open(out_file, "r", encoding="utf-8") as out:
+            assert out.read() == ""
+        with open(err_file, "r", encoding="utf-8") as err:
+            content = err.read()
+            assert "unrecognized option" in content
+
+
+def test_popen_issues_unique_ids(shell_launcher, test_dir):
+    """Validate that all ids are unique within ShellLauncher._launched"""
+    run_dir, out_file, err_file = generate_directory(test_dir)
+    with (
+        open(out_file, "w", encoding="utf-8") as out,
+        open(err_file, "w", encoding="utf-8") as err,
+    ):
+        cmd = ShellLauncherCommand(
+            {}, run_dir, out, err, EchoHelloWorldEntity().as_program_arguments()
+        )
+        for _ in range(5):
+            _ = shell_launcher.start(cmd)
+        assert len(shell_launcher._launched) == 5
+        assert all(proc.wait() == 0 for proc in shell_launcher._launched.values())
+
```

Comment:
test title does not match the impl: Consider adding a check for id uniqueness when calling `start`
```py
seen = set()
for _ in range(5):
    id = shell_launcher.start(cmd)
    assert id not in seen, "Duplicate ID issued"
    seen.add(id)
```

Reply:
ah I see